databases/core.py

```diff
@@ -385,31 +388,38 @@ async def wrapper(*args: typing.Any, **kwargs: typing.Any) -> typing.Any:
         return wrapper  # type: ignore
 
     async def start(self) -> "Transaction":
-        self._connection = self._connection_callable()
-        self._transaction = self._connection._connection.transaction()
-
-        async with self._connection._transaction_lock:
-            is_root = not self._connection._transaction_stack
-            await self._connection.__aenter__()
-            await self._transaction.start(
-                is_root=is_root, extra_options=self._extra_options
-            )
-            self._connection._transaction_stack.append(self)
+        connection = self._connection_callable()
+        transaction = connection._connection.transaction()
+        self._transaction_context.set(transaction)
```

Comment:
Technically you should probably stash the "token" this returns and reset the variable via the token instead of setting it to `None` later, but I'm not sure if there would ever be nesting in normal usage.

Reply:
Yeah, that's part of the contextvars library I don't really understand. The current implementation for connections doesn't do this right now either, though. Worse, it was assigning a newly created ContextVar to clear the state of the connection.

I have to stash the token somehow, but since we're not immediately restoring the value, I think I'd need yet another context variable to keep the reset token! I can't just use an instance property, for the same reason that this PR exists, either, so where does it end? It seems that the API is designed only for cases like this: 
```python
from contextvars import ContextVar

var = ContextVar('some_var')

async def some_library_function():
    token = var.set(42)
    try: 
        await _do_something()
    finally:
        var.reset(token)
```

Since databases has separate time-of-creation and time-of-use for these context vars (eg `Transaction.start()`, versus `Transaction.commit()` or `Transaction.rollback()`), I don't see a clear way to keep the token around - which is why I opted for setting the value to None instead. Databases is not using these ContextVars to interface with other codebases, so as Databases is the sole owner, I think setting to None is fine.

Reply:
Yeah, neither this PR nor the current master is safe in this regard. I explored the idea of a contextvar containing a dict and I have something that passes all tests now, but I need to also confirm with an MCVE as well / add some tests since the tests passed before this PR anyway.

Reply:
I switched to an implementation where the two contextvars each store a dictionary for the database connections or for the active transactions. We should now be safe in all of these cases: 

- The simple case where a Database is instantiated and used immediately - it gets its own connection
- A Database is instantiated and used in closures that run in the same tasks - the same connection is reused
- A Database is instantiated and used in closures that run in one or many descendant tasks - multiple connections are created, one per task.
- Multiple Databases are instantiated and used in the same task, eg via nesting `with` statements - each database instance uses and reuses a connection unique to itself
- Multiple Databases are instantiated and used in separate descendant tasks - each database instance uses its own connection
- In all the above cases, each Transaction instance is always paired with the connection used in the combination of the current task and parent database. The Transaction instance ID is used when multiple nested connections are used in the same task.
- This implementation now ensures a task-local mapping of transaction to transactionbackend, and support for nested nested transactions, including when a child transaction is run in a separate task.

Reply:
Resolved: @madkinsz took a new approach that I think is best here - back to instance-local storage, but without context variables, and with dictionaries keyed on the current asyncio task